lib/bing-ads-reporting/core_service.rb

```diff
@@ -0,0 +1,158 @@
+    end
+
+    def poll_report(id)
+      begin
+        client.call(poll_operation, message: generate_poll_message(id))
+      rescue Savon::SOAPFault => e
+        err = e.to_hash[:fault][:detail][:ad_api_fault_detail][:errors][:ad_api_error][:error_code] rescue nil
+        msg = e.to_hash[:fault][:detail][:ad_api_fault_detail][:errors][:ad_api_error][:message] if err
+        if err.nil?
+          err = e.to_hash[:fault][:detail][:api_fault_detail][:operation_errors][:operation_error][:error_code] rescue nil
+          msg = e.to_hash[:fault][:detail][:api_fault_detail][:operation_errors][:operation_error][:message] if err
+        end
+        if err == 'AuthenticationTokenExpired'
+          @logger.error err
+          raise AuthenticationTokenExpired.new(msg)
+        end
+      end
+    end
+
+    def report_url(id)
+      polled = poll_report(id)
+      status = get_status(polled.body)
+      download_url = get_download_url(polled.body)
+      return nil if download_url.nil? && status == success_status
+      raise "Report URL is not available for report id #{id}" unless download_url
+      download_url
+    end
+
+    def download(url)
+      return unless url
+      @logger.debug "Downloading Bing report from: #{url}"
+      curl = Curl::Easy.new(url)
+      curl.perform
+      curl.body_str
+    end
+
+    def client
+      if @settings[:username] && @settings[:password]
+        header = {ns('ApplicationToken') => @settings[:applicationToken],
+                                            ns('CustomerAccountId') => @settings[:accountId],
+                                            ns('CustomerId') => @settings[:customerId],
+                                            ns('DeveloperToken') => @settings[:developerToken],
+                                            ns('UserName') => @settings[:username],
+                                            ns('Password') => @settings[:password] }
+      else
+        header = {ns('ApplicationToken') => @settings[:applicationToken],
+                                            ns('CustomerAccountId') => @settings[:accountId],
+                                            ns('CustomerId') => @settings[:customerId],
+                                            ns('DeveloperToken') => @settings[:developerToken],
+                                            ns('AuthenticationToken') => @settings[:authenticationToken] }
+      end
+      Savon.client({
+                    wsdl: wdsl,
+                    log_level: :info,
+                    namespaces: { 'xmlns:arr' => 'http://schemas.microsoft.com/2003/10/Serialization/Arrays',
+                                  'xmlns:i' => "http://www.w3.org/2001/XMLSchema-instance" },
+                    soap_header: header
+                    })
+                    # .merge({pretty_print_xml: true, log_level: :debug, log: true, logger: @logger})) # for more logging
```

Comment:
If we don't need that, can we remove it please?

Reply:
it comes in handy in tough times ;)

Reply:
yeah agreed but I'd say remove it or leave it and un-comment as it's logging in debug anyway

Reply:
removed